app: Centralize directory validation

stephenfin opened the pull request:
This allows us to avoid duplication of code and ensure validation happens regardless of who's initializing the class.

stephenfin committed:
app: Centralize directory validation

This allows us to avoid duplication of code and ensure validation
happens regardless of who's initializing the class. We introduce a new
exception - ApplicationError - to indicate these kinds of issues. This
subclasses SphinxError, meaning we don't need to modify our exception
handling code.

Signed-off-by: Stephen Finucane <stephen@that.guru>

tk0miya commented on sphinx/application.py:
```diff
@@ -123,10 +124,24 @@ def __init__(self, srcdir, confdir, outdir, doctreedir, buildername,
+        # validate provided directories
+        self.srcdir = abspath(srcdir)
+        self.outdir = abspath(outdir)
+        self.doctreedir = abspath(doctreedir or path.join(outdir, '.doctrees'))
```
IMO, it is better that `Sphinx` class simply uses given `doctreedir` argument.

stephenfin replied:
OK. I think my reasoning for doing this was that the pattern was duplicated in two places. That's probably not a bad thing though. Perhaps I should raise an exception if these are unset?

tk0miya replied:
Yes, I think so. I think new responsibility of `Sphinx` class is regulating paths and checking their existence (validation). So exceptions should be raised if invalid parameter given.

I think reducing code duplication is not important. IMO, it is more important to clear responsibility of each modules. Before, `Sphinx` class and its clients are strongly coupled. `Sphinx` always trusts arguments from clients and does not check them.
With this change, `Sphinx` class becomes defensive. It also means less coupled.

Note: I think choosing `doctreedir` is a responsibility of clients. It's simple.

stephenfin replied:
Done

tk0miya commented on sphinx/application.py:
```diff
@@ -123,10 +124,24 @@ def __init__(self, srcdir, confdir, outdir, doctreedir, buildername,
+                                  "conf.py file (%s)" % confdir)
+
+        if not path.isdir(self.srcdir):
+            raise ConfigError('Cannot find source directory (%s)' %
```
+1 for checking directory existence and regulate abspath at `Sphinx` class.
But I feel `ConfigError` is not good for this case. It is used for configuration error.
So these validation should raise other error.

stephenfin replied:
This is different to the issue below. IMO, the source directory is part of the configuration. For example, calling `sphinx-build SRCDIR OUTDIR` *configures* Sphinx to read input from `SRCDIR` and write output to `OUTDIR`. However, as noted below, perhaps you think that `ConfigError` is only for configuration file errors? If so, perhaps I should add a new `ConfigError` exception and rename the existing one `ConfigFileError`? The renamed exception can subclass the new one.

tk0miya replied:
>However, as noted below, perhaps you think that ConfigError is only for configuration file errors?

Yes. I understand the command line arguments are part of configuration in a broad sense. But , in narrow sense, it is not. In a broad sense, almost of user inputs are "configuration". But it is too ambiguous.

To be clearly, I propose you to add new exception like `ArgumentError`. I prefer to use different name for these errors.

stephenfin replied:
Done

tk0miya commented on sphinx/setup_command.py:
```diff
@@ -138,20 +138,14 @@ def finalize_options(self):
             self.mkpath(self.build_dir)  # type: ignore
-        self.build_dir = abspath(self.build_dir)
         self.doctree_dir = os.path.join(self.build_dir, 'doctrees')
+        # TODO(stephenfin): Is this necessary?
```
It is not needed. Builder class creates doctreedir if not exist.
https://github.com/sphinx-doc/sphinx/blob/a1f13e30000cdeceef7fc29b2bf3d0f45d5750ca/sphinx/builders/__init__.py#L85

stephenfin replied:
I'll fix this in the next version

stephenfin replied:
Done

tk0miya commented on sphinx/util/osutil.py:
```diff
@@ -219,7 +221,11 @@ def abspath(pathdir):
+        try:
+            pathdir = pathdir.decode(fs_encoding)
+        except UnicodeError:
+            raise ConfigError('multibyte filename not supported on this '
```
I must say it is very strange that `abspath()` raises `ConfigError`.

stephenfin replied:
@tk0miya I used this because it seemed like a configuration issue. However, it is not a configuration *file* issue. What would you suggest instead? Remember that we'll raise this exception for any and all issues with provided paths.

tk0miya replied:
No, this is not related with case of `Sphinx` class. this `abspath()` is an utility function. So it is used for other paths. So I think using standard exception is better. So I think it is better to raise `UnicodeError` directly.

stephenfin replied:
Done

tk0miya reviewed (changes requested):
I agree with basic concept of this PR.

stephenfin committed:
setuptools: Don't create directories

Don't create the build or doctree directories. These are already handled
by 'sphinx.application.Application' and 'sphinx.builder.Builder',
respectively.

Signed-off-by: Stephen Finucane <stephen@that.guru>

tk0miya commented:
Thank you for update. LGTM!
